Source/Core/Core/WatchDog.h

```diff
@@ -0,0 +1,113 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include "Common/CommonTypes.h"
+
+#include <compare>
+#include <deque>
+#include <mutex>
+#include <string_view>
+#include <unordered_map>
+#include <vector>
+
+#include <fmt/format.h>
+
+namespace WatchDog
+{
+using Marker = std::string_view;
+
+struct TimeSlice
+{
+public:
+  TimeSlice(const Marker& marker = Marker(), DT duration = DT::zero())
+      : marker{marker}, duration{duration}
+  {
+  }
+
+  TimeSlice(TimeSlice&&) = default;
+  TimeSlice(const TimeSlice&) = default;
+  TimeSlice& operator=(TimeSlice&&) = default;
+  TimeSlice& operator=(const TimeSlice&) = default;
+
+  Marker marker;
+  DT duration;
+};
+
+struct ProfilerEntry
+{
+public:
+  ProfilerEntry(const Marker& name, const DT duration, const DT total_duration)
+      : name{name}, percentage{DT_s(duration) / DT_s(total_duration)}, duration{duration}
+  {
+  }
+
+  ProfilerEntry(ProfilerEntry&) = default;
+  ProfilerEntry(const ProfilerEntry&) = default;
+  ProfilerEntry& operator=(ProfilerEntry&&) = default;
+  ProfilerEntry& operator=(const ProfilerEntry&) = default;
+
+  std::string GetString() const { return fmt::format("{:5.2f}% - {}", 100.0 * percentage, name); }
+
+  friend auto operator<=>(const ProfilerEntry& a, const ProfilerEntry& b)
+  {
+    return a.percentage <=> b.percentage;
+  };
+
+  Marker name;
+  double percentage;
+  DT duration;
+};
+
+class Profiler
+{
+public:
+  Profiler() { Reset(); };
+
+  Profiler(const Profiler&) = delete;
+  Profiler& operator=(const Profiler&) = delete;
+
+  void Reset(const Marker& first_marker = Marker("Nothing"));
+
+  void SetMarker(const Marker& next_marker);
+
+  void PushMarker(const Marker& next_marker);
```

Comment:
I'd suggest (but pick a better name?):

`ScopedMarker TempMarker(const Marker& next_marker);`

Make `ScopedMarker` an internal class and move `PushMarker` and `PopMarker` to be private functions.  Every single use case (thus far?) in the usage seems to be a scoped case.

Reply:
Ah `ScopedMarker` is also used outside `Profiler` (in VertexManager) so I guess it can't be internal.  Maybe it could be used more often then?